File: agent-service/src/moeazi_agent/model_routing.py
from dataclasses import replace
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .roles import Assignment


class OpenRouterPreferences(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)
    sort: str = "price"
    allow_fallbacks: bool = Field(default=True, alias="allowFallbacks")
    allowed_providers: list[str] = Field(default_factory=list, alias="allowedProviders", max_length=20)
    ignored_providers: list[str] = Field(default_factory=list, alias="ignoredProviders", max_length=20)
    data_collection: str = Field(default="deny", alias="dataCollection")
    zero_data_retention: bool = Field(default=True, alias="zeroDataRetention")

    @model_validator(mode="after")
    def validate_preferences(self):
        if self.sort not in {"price", "latency", "throughput"}:
            raise ValueError("Unsupported OpenRouter sort")
        if self.data_collection not in {"allow", "deny"}:
            raise ValueError("Unsupported OpenRouter data policy")
        if self.allowed_providers and self.ignored_providers:
            overlap = set(self.allowed_providers) & set(self.ignored_providers)
            if overlap:
                raise ValueError("OpenRouter providers cannot be both allowed and ignored")
        providers = self.allowed_providers + self.ignored_providers
        if any(
            not item or len(item) > 80
            or any(character not in "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._/-" for character in item)
            for item in providers
        ):
            raise ValueError("Invalid OpenRouter provider slug")
        return self


class ModelRoute(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)
    slot: str
    provider: str
    model: str
    credential_source: str = Field(alias="credentialSource")
    connection_id: str | None = Field(default=None, alias="connectionId")
    max_output_tokens: int = Field(alias="maxOutputTokens", ge=128, le=8_192)
    reasoning_effort: str | None = Field(default=None, alias="reasoningEffort")
    input_price: int = Field(alias="inputPriceMicrousdPerMillion", ge=0)
    cached_input_price: int = Field(alias="cachedInputPriceMicrousdPerMillion", ge=0)
    output_price: int = Field(alias="outputPriceMicrousdPerMillion", ge=0)
    estimated_input_price: int | None = Field(
        default=None, alias="estimatedInputPriceMicrousdPerMillion", ge=0,
    )
    estimated_output_price: int | None = Field(
        default=None, alias="estimatedOutputPriceMicrousdPerMillion", ge=0,
    )
    openrouter: OpenRouterPreferences | None = None


class ModelRouting(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)
    schema_version: int = Field(alias="schemaVersion")
    routes: list[ModelRoute]

    @model_validator(mode="after")
    def validate_routes(self):
        if self.schema_version not in {1, 2}:
            raise ValueError("Unsupported model-routing schema")
        slots = {item.slot for item in self.routes}
        if not {"specialist_default", "synthesis"}.issubset(slots):
            raise ValueError("Specialist and synthesis routes are required")
        credentials: dict[str, str] = {}
        for route in self.routes:
            if route.provider not in {"openai", "deepseek", "openrouter"}:
                raise ValueError("Unsupported provider")
            if route.credential_source not in {"platform", "byok"}:
                raise ValueError("Unsupported credential source")
            if route.credential_source == "byok" and not route.connection_id:
                raise ValueError("BYOK route requires a connection")
            if route.provider == "openrouter":
                if route.credential_source != "byok":
                    raise ValueError("OpenRouter routes are BYOK only")
                if route.openrouter is None:
                    route.openrouter = OpenRouterPreferences()
            elif route.openrouter is not None:
                raise ValueError("OpenRouter preferences require the OpenRouter provider")
            key = f"{route.credential_source}:{route.connection_id or 'platform'}"
            if route.provider in credentials and credentials[route.provider] != key:
                raise ValueError("Each provider must use one credential per configuration")
            credentials[route.provider] = key
        return self

    def route(self, slot: str) -> ModelRoute:
        exact = next((item for item in self.routes if item.slot == slot), None)
        legacy = {
            "synthesis_primary": "openai_synthesis",
            "synthesis_challenger": "deepseek_synthesis",
        }.get(slot)
        if not exact and legacy:
            exact = next((item for item in self.routes if item.slot == legacy), None)
        if exact:
            return exact
        fallback = "specialist_default" if slot.startswith("role:") else "synthesis"
        return next(item for item in self.routes if item.slot == fallback)


def provider_preferences(route: ModelRoute) -> dict | None:
    if not route.openrouter:
        return None
    preferences = route.openrouter.model_dump(by_alias=True)
    preferences["maxInputPriceMicrousdPerMillion"] = route.input_price
    preferences["maxOutputPriceMicrousdPerMillion"] = route.output_price
    return preferences


def routed_assignments(assignments: list[Assignment], routing: ModelRouting | None) -> list[Assignment]:
    if not routing:
        return assignments
    return [
        replace(
            item,
            provider=routing.route(f"role:{item.role.name}").provider,
            model=routing.route(f"role:{item.role.name}").model,
            credential_source=routing.route(f"role:{item.role.name}").credential_source,
            max_output_tokens=routing.route(f"role:{item.role.name}").max_output_tokens,
            reasoning_effort=routing.route(f"role:{item.role.name}").reasoning_effort,
            provider_preferences=provider_preferences(routing.route(f"role:{item.role.name}")),
        )
        for item in assignments
    ]


def call_cost(route: ModelRoute, usage: Any) -> int:
    if route.credential_source != "byok":
        return 0
    reported = getattr(usage, "provider_cost_microusd", None)
    if reported is not None:
        return max(0, int(reported))
    cached = min(max(0, usage.cached_input_tokens), max(0, usage.input_tokens))
    uncached = max(0, usage.input_tokens - cached)
    return round((
        uncached * route.input_price + cached * route.cached_input_price
        + max(0, usage.output_tokens) * route.output_price
    ) / 1_000_000)


def call_credits(route: ModelRoute, kind: str) -> int:
    if route.credential_source == "platform":
        return 3 if kind == "specialist" else 9 if kind == "arbiter" else 7
    return 1 if kind == "specialist" else 3 if kind == "arbiter" else 2


def route_estimate(tier: str, assignments: list[Assignment], routing: ModelRouting | None) -> dict:
    if not routing:
        from .credits import estimated_credits
        return {"credits": estimated_credits(tier), "provider_cost_microusd": 0}
    specialist = routed_assignments(assignments, routing)
    steps = {
        "focus": ["synthesis"],
        "pro": ["critic", "synthesis"],
        "max": ["synthesis_primary", "synthesis_challenger", "arbiter"],
    }[tier]
    calls = [(routing.route(f"role:{item.role.name}"), "specialist") for item in specialist]
    calls.extend((routing.route(step), "arbiter" if step == "arbiter" else "synthesis") for step in steps)
    estimated = sum(
        round((
            (route.estimated_input_price if route.estimated_input_price is not None else route.input_price) * 2_500
            + (route.estimated_output_price if route.estimated_output_price is not None else route.output_price)
            * route.max_output_tokens
        ) / 1_000_000)
        for route, _ in calls if route.credential_source == "byok"
    )
    maximum = sum(
        round((route.input_price * 2_500 + route.output_price * route.max_output_tokens) / 1_000_000)
        for route, _ in calls if route.credential_source == "byok"
    )
    return {
        "credits": sum(call_credits(route, kind) for route, kind in calls),
        "provider_cost_microusd": maximum,
        "estimated_provider_cost_microusd": estimated,
    }

File: convex/helpers/modelRouting.ts
const ROLES = [
  "technical_trend", "liquidity", "derivatives", "on_chain", "news", "social",
  "cross_venue_basis", "volatility_liquidations", "whale_flow", "macro_correlation",
  "execution_quality", "portfolio_exposure", "short_horizon", "swing_horizon",
  "bull_case", "bear_case", "range_regime", "catalyst", "data_quality_skeptic",
  "market_integrity",
] as const;

const SYNTHESIS = {
  focus: ["synthesis"],
  pro: ["critic", "synthesis"],
  max: ["synthesis_primary", "synthesis_challenger", "arbiter"],
} as const;

export type OpenRouterPreferences = {
  sort: "price" | "latency" | "throughput";
  allowFallbacks: boolean;
  allowedProviders: string[];
  ignoredProviders: string[];
  dataCollection: "allow" | "deny";
  zeroDataRetention: boolean;
};

export type ModelRoute = {
  slot: string;
  provider: "openai" | "deepseek" | "openrouter";
  model: string;
  credentialSource: "platform" | "byok";
  connectionId?: string;
  maxOutputTokens: number;
  reasoningEffort?: "none" | "low" | "medium" | "high";
  inputPriceMicrousdPerMillion: number;
  cachedInputPriceMicrousdPerMillion: number;
  outputPriceMicrousdPerMillion: number;
  estimatedInputPriceMicrousdPerMillion?: number;
  estimatedOutputPriceMicrousdPerMillion?: number;
  openrouter?: OpenRouterPreferences;
};

export type ModelRoutingDocument = {
  schemaVersion: 2;
  routes: ModelRoute[];
};

const platformRoute = (
  slot: string, provider: "openai" | "deepseek", model: string, maxOutputTokens: number,
): ModelRoute => ({
  slot, provider, model, credentialSource: "platform", maxOutputTokens,
  inputPriceMicrousdPerMillion: 0, cachedInputPriceMicrousdPerMillion: 0,
  outputPriceMicrousdPerMillion: 0,
});

export function defaultModelRouting(): ModelRoutingDocument {
  return {
    schemaVersion: 2,
    routes: [
      platformRoute("specialist_default", "deepseek", "deepseek-v4-flash", 700),
      platformRoute("synthesis", "openai", "gpt-5.6-sol", 1_200),
      platformRoute("critic", "deepseek", "deepseek-v4-pro", 1_200),
      platformRoute("arbiter", "openai", "gpt-5.6-sol", 1_200),
      platformRoute("synthesis_primary", "openai", "gpt-5.6-sol", 1_200),
      platformRoute("synthesis_challenger", "deepseek", "deepseek-v4-pro", 1_200),
      platformRoute("policy_draft", "deepseek", "deepseek-v4-flash", 700),
    ],
  };
}

const ALLOWED_DEFAULTS = new Set([
  "specialist_default", "synthesis", "critic", "arbiter", "policy_draft",
  "synthesis_primary", "synthesis_challenger", "openai_synthesis", "deepseek_synthesis",
]);

function validSlot(slot: string) {
  return ALLOWED_DEFAULTS.has(slot)
    || (slot.startsWith("role:") && ROLES.includes(slot.slice(5) as typeof ROLES[number]));
}

export function parseModelRouting(value: string): ModelRoutingDocument {
  const input = JSON.parse(value) as { schemaVersion: number; routes: ModelRoute[] };
  if (![1, 2].includes(input?.schemaVersion) || !Array.isArray(input.routes)) {
    throw new Error("Unsupported model-routing document.");
  }
  const parsed: ModelRoutingDocument = {
    schemaVersion: 2,
    routes: input.routes.map((route) => ({
      ...route,
      slot: route.slot === "openai_synthesis" ? "synthesis_primary"
        : route.slot === "deepseek_synthesis" ? "synthesis_challenger" : route.slot,
      openrouter: route.provider === "openrouter" ? {
        sort: route.openrouter?.sort ?? "price",
        allowFallbacks: route.openrouter?.allowFallbacks ?? true,
        allowedProviders: route.openrouter?.allowedProviders ?? [],
        ignoredProviders: route.openrouter?.ignoredProviders ?? [],
        dataCollection: route.openrouter?.dataCollection ?? "deny",
        zeroDataRetention: route.openrouter?.zeroDataRetention ?? true,
      } : undefined,
    })),
  };
  if (parsed.routes.length < 2 || parsed.routes.length > 32) {
    throw new Error("Model routing must contain 2-32 routes.");
  }
  const slots = new Set<string>();
  const credentials = new Map<string, string>();
  for (const route of parsed.routes) {
    if (!validSlot(route.slot) || slots.has(route.slot)) throw new Error(`Invalid or duplicate route: ${route.slot}`);
    slots.add(route.slot);
    if (!["openai", "deepseek", "openrouter"].includes(route.provider)) throw new Error("Unsupported model provider.");
    if (!["platform", "byok"].includes(route.credentialSource)) throw new Error("Unsupported credential source.");
    if (!/^[A-Za-z0-9][A-Za-z0-9._:/-]{0,127}$/.test(route.model)) throw new Error("Invalid model ID.");
    if (route.credentialSource === "byok" && !route.connectionId) throw new Error("BYOK routes require a connection.");
    if (route.provider === "openrouter" && route.credentialSource !== "byok") {
      throw new Error("OpenRouter routes are BYOK only.");
    }
    if (route.provider !== "openrouter" && route.openrouter) {
      throw new Error("OpenRouter preferences require an OpenRouter route.");
    }
    if (route.openrouter) validateOpenRouter(route.openrouter);
    const credential = `${route.credentialSource}:${route.connectionId ?? "platform"}`;
    const current = credentials.get(route.provider);
    if (current && current !== credential) throw new Error("Each provider must use one credential source per configuration.");
    credentials.set(route.provider, credential);
    if (route.maxOutputTokens < 128 || route.maxOutputTokens > 8_192) throw new Error("Invalid output-token cap.");
    for (const price of [
      route.inputPriceMicrousdPerMillion, route.cachedInputPriceMicrousdPerMillion,
      route.outputPriceMicrousdPerMillion, route.estimatedInputPriceMicrousdPerMillion ?? 0,
      route.estimatedOutputPriceMicrousdPerMillion ?? 0,
    ]) {
      if (!Number.isFinite(price) || price < 0 || price > 1_000_000_000) throw new Error("Invalid model rate.");
    }
    if (route.credentialSource === "byok" && route.inputPriceMicrousdPerMillion === 0 && route.outputPriceMicrousdPerMillion === 0) {
      throw new Error(`BYOK route ${route.slot} requires confirmed model rates.`);
    }
  }
  if (!slots.has("specialist_default") || !slots.has("synthesis")) {
    throw new Error("Specialist and synthesis defaults are required.");
  }
  return parsed;
}

function validateOpenRouter(value: OpenRouterPreferences) {
  if (!["price", "latency", "throughput"].includes(value.sort)) throw new Error("Invalid OpenRouter sort.");
  if (!["allow", "deny"].includes(value.dataCollection)) throw new Error("Invalid OpenRouter data policy.");
  const overlap = value.allowedProviders.filter((item) => value.ignoredProviders.includes(item));
  if (overlap.length) throw new Error("An upstream provider cannot be both allowed and ignored.");
  const providers = [...value.allowedProviders, ...value.ignoredProviders];
  if (providers.length > 40 || providers.some((item) => !/^[A-Za-z0-9][A-Za-z0-9._/-]{0,79}$/.test(item))) {
    throw new Error("Invalid OpenRouter upstream provider.");
  }
}

function routeFor(routes: ModelRoute[], slot: string) {
  return routes.find((item) => item.slot === slot)
    ?? routes.find((item) => item.slot === (slot.startsWith("role:") ? "specialist_default" : "synthesis"))!;
}

export function estimateModelRouting(document: ModelRoutingDocument, tier: "focus" | "pro" | "max") {
  const base = ROLES.slice(0, { focus: 6, pro: 12, max: 20 }[tier]);
  const roles = tier === "pro"
    ? [...base, "liquidity", "derivatives", "cross_venue_basis", "volatility_liquidations"]
    : tier === "max" ? [...base, ...base.filter((_, index) => [0, 1, 2, 6, 7, 8, 11, 12, 13, 14].includes(index))] : base;
  const calls = [
    ...roles.map((role) => ({ route: routeFor(document.routes, `role:${role}`), kind: "specialist" })),
    ...SYNTHESIS[tier].map((step) => ({ route: routeFor(document.routes, step), kind: step === "arbiter" ? "arbiter" : "synthesis" })),
  ];
  const credits = calls.reduce((sum, call) => {
    if (call.route.credentialSource === "platform") return sum + (call.kind === "specialist" ? 3 : call.kind === "arbiter" ? 9 : 7);
    return sum + (call.kind === "specialist" ? 1 : call.kind === "arbiter" ? 3 : 2);
  }, 0);
  const cost = (route: ModelRoute, maximum: boolean) => {
    if (route.credentialSource !== "byok") return 0;
    const input = maximum ? route.inputPriceMicrousdPerMillion
      : route.estimatedInputPriceMicrousdPerMillion ?? route.inputPriceMicrousdPerMillion;
    const output = maximum ? route.outputPriceMicrousdPerMillion
      : route.estimatedOutputPriceMicrousdPerMillion ?? route.outputPriceMicrousdPerMillion;
    return Math.round(input * 2_500 / 1_000_000)
      + Math.round(output * route.maxOutputTokens / 1_000_000);
  };
  return {
    calls: calls.length, credits,
    estimatedProviderCostMicrousd: calls.reduce((sum, call) => sum + cost(call.route, false), 0),
    maximumProviderCostMicrousd: calls.reduce((sum, call) => sum + cost(call.route, true), 0),
  };
}
